tests/script/core/test_validation.py

```diff
@@ -0,0 +1,119 @@
+from typing import Dict
+from dff.pipeline import Pipeline
+from dff.script import PRE_RESPONSE_PROCESSING, PRE_TRANSITIONS_PROCESSING, RESPONSE, TRANSITIONS, Context, Message, Script, NodeLabel3Type
+from dff.script.conditions import exact_match
+from dff.script.labels import repeat
+
+
+def wrong_param_number(number: int) -> float:
+    return 8.0 + number
+
+
+def wrong_param_types(number: int, flag: bool) -> float:
+    return 8.0 + number if flag else 42.1
+
+
+def wrong_return_type(_: Context, __: Pipeline) -> float:
+    return 1.0
+
+
+def correct_label(_: Context, __: Pipeline) -> NodeLabel3Type:
+    return ("root", "start", 1)
+
+
+def correct_response(_: Context, __: Pipeline) -> Message:
+    return Message("hi")
+
+
+def correct_condition(_: Context, __: Pipeline) -> bool:
+    return True
+
+
+def correct_pre_response_processor(_: Context, __: Pipeline) -> None:
+    pass
+
+
+def correct_pre_transition_processor(_: Context, __: Pipeline) -> None:
+    pass
+
+
+def function_signature_test(message: str, param_number: Dict, param_types: Dict, return_type: Dict):
+    for script, error in map(["param number", "param types", "return type"], [param_number, param_types, return_type]):
+        try:
+            Script(script=script)
+            raise Exception(f"can not be passed: {message}: {error}")
+        except ValueError:
+            pass
+
+
+def test_labels():
+    param_number_script = {"root": { "start": { TRANSITIONS: { wrong_param_number: exact_match(Message("hi")) } } } }
+    param_types_script = {"root": { "start": { TRANSITIONS: { wrong_param_types: exact_match(Message("hi")) } } } }
+    return_type_script = {"root": { "start": { TRANSITIONS: { wrong_return_type: exact_match(Message("hi")) } } } }
+    function_signature_test("wrong label function signature", param_number_script, param_types_script, return_type_script)
+    try:
+        # wrong label tuple flow
+        Script(script={"root": { "start": { TRANSITIONS: { ("other", "start", 1): exact_match(Message("hi")) } } } })
+        raise Exception("can not be passed: wrong label tuple flow")
+    except ValueError:
```

Comment:
This code is unrelated to others in the same test suite.
Group multiple tests testing the same part into a class:
https://docs.pytest.org/en/7.1.x/getting-started.html#group-multiple-tests-in-a-class
Each method should test a separate case.
That way the cases become independent.

Reply:
Yes.
Why wouldn't we want to use test classes? If the only reason is that we haven't used them before then we should start doing so.
They help with test case organization among other things.

> This code is actually related

What I meant is that these test cases are independent from each other (one is not a requirement for another).
Failing the first one would lead to us not even testing any of the other ones.

By creating a class `TestLabelValidation` we group the test cases in the intended pytest way -- they are still grouped but now they are executed separately.